### PR TITLE
[Feature] idmapper map precursors

### PR DESCRIPTION
**Pull request by timosachsenberg**
in many cases it makes sense to map unidentified precursors to consensus features.
This PR proposes a mechanism that annotates a spectrum_index (intList) user parameter to consensus features.

### CODE REVIEW

**Comment by hendrikweisser** [comment]
Hey Timo, why only consensus features? What about simple features (featureXML)?

**Comment by timosachsenberg** [comment]
Definitly nice to have. I am just currently prioritizing stuff we need in to build the basis for protein-protein cross-linking functionality into OpenMS. (e.g. in the xQuest algorithm you align MS2, mapped to consensus features from light and heavy pairs before doing the actual identification)

File: src/openms/source/ANALYSIS/ID/IDMapper.cpp [hunk 1]
```diff
@@ -229,6 +229,146 @@ namespace OpenMS
...
+
+          if (was_added)
+            break;
+
+        } // m/z values to check
+
+        // break to here
+
+      } // features
+    } // Identifications
```
Comment (mwalzer) [inline src/openms/source/ANALYSIS/ID/IDMapper.cpp]: Did I miss the identifications?

**Comment by mwalzer** [comment]
Definitely it would be nice to have annotated, which features (in featuremap or consensusmap) do have precursors but no identifications

File: src/openms/source/ANALYSIS/ID/IDMapper.cpp (hunk 1 above)
Comment (timosachsenberg) [inline src/openms/source/ANALYSIS/ID/IDMapper.cpp]: ?

Comment (mwalzer) [inline src/openms/source/ANALYSIS/ID/IDMapper.cpp]:
```
} // Identifications
```
But right now you only tag precursors to the consensus feature so no identifications? Is that planned?

Comment (hendrikweisser) [inline src/openms/source/ANALYSIS/ID/IDMapper.cpp]: There should already be a separate function for annotating a consensus map with identifications. Maybe the `// Identifications` comment is a mistake and should be `// precursors`?

Comment (timosachsenberg) [inline src/openms/source/ANALYSIS/ID/IDMapper.cpp]: The comment is probably a copy and paste error. It should say precursors.

**Comment by timosachsenberg** [comment]
@hendrikweisser and @mwalzer I now annotate both specta and native id on the PeptideIdentification for each feature. After linking these are carried over to the consensus map together with the "map index" it mapped to.

**Comment by mwalzer** [comment]
:+1:

**Comment by timosachsenberg** [comment]
ok I will simplify a bit and push again

File: src/openms/source/ANALYSIS/ID/IDMapper.cpp [hunk 2]
```diff
@@ -190,253 +200,176 @@ namespace OpenMS
...
+        PeptideIdentification precursor_empty_id;
+        precursor_empty_id.setRT(rt_value);
+        precursor_empty_id.setMZ(mz_p);
+        precursor_empty_id.setMetaValue("spectrum_index", spectrum_index);
+        if (!spectra[spectrum_index].getNativeID().empty())
         {
-          double mz_prec = precursors[i_p].getMZ();
+          precursor_empty_id.setMetaValue("spectrum_reference",  spectra[spectrum_index].getNativeID());
+        }
+        precursor_empty_id.setIdentifier(empty_protein_id.getIdentifier());
```
Comment (mwalzer) [inline src/openms/source/ANALYSIS/ID/IDMapper.cpp]: :+1:

File: src/openms/source/ANALYSIS/ID/IDMapper.cpp [hunk 3]
```diff
@@ -190,253 +200,176 @@ namespace OpenMS
...
-        // iterate over m/z values of precursors
-        for (Size i_p = 0; i_p < precursors.size(); ++i_p)
+        PeptideIdentification precursor_empty_id;
+        precursor_empty_id.setRT(rt_value);
+        precursor_empty_id.setMZ(mz_p);
+        precursor_empty_id.setMetaValue("spectrum_index", spectrum_index);
+        if (!spectra[spectrum_index].getNativeID().empty())
         {
-          double mz_prec = precursors[i_p].getMZ();
+          precursor_empty_id.setMetaValue("spectrum_reference",  spectra[spectrum_index].getNativeID());
```
Comment (mwalzer) [inline src/openms/source/ANALYSIS/ID/IDMapper.cpp]: maybe warn if there is no NativeID?

Comment (timosachsenberg) [inline src/openms/source/ANALYSIS/ID/IDMapper.cpp]: I think we should warn in the file handlers. Warning in other classes might be possible but then you'll get potentially a lot of warnings.

**Comment by timosachsenberg** [comment]
@mwalzer any further comments?

**Pull request merged**